lib/wit/gitrepo.py

```diff
@@ -242,14 +246,91 @@ def is_ancestor(self, ancestor, current=None):
 
+    def _read_submodules_from_commit(self, revision) -> List[RepoEntry]:
+        proc = self._git_command("show", "{}:{}".format(revision, GitRepo.SUBMODULE_FILE))
+        if proc.returncode:
+            log.debug("No .gitmodule file found in repo [{}:{}]".format(revision, self.path))
+            return []
+
+        log.debug("{}:{} does not have wit-manifest.json, "
+                  "reading dependencies from git submodules instead"
+                  .format(self.name, revision))
+
+        # Use the 'git config' parser to read the submodule contents.
+        # Output is of the form:
+        #     submodule.$NAME.path $PATH
+        #     submodule.$NAME.url  $REMOTE
+        proc = self._git_command("config", "-f-", "--get-regex", "submodule..*",
+                                 input=proc.stdout)
+        self._git_check(proc)
+
+        paths_by_name = {}
+        urls_by_name = {}
+
+        path_r = re.compile('^submodule.(.*).path (.*)$')
+        for line in proc.stdout.splitlines():
+            m = path_r.match(line)
+            if m:
+                paths_by_name[m.group(1)] = m.group(2)
+
+        url_r = re.compile('^submodule.(.*).url (.*)$')
+        for line in proc.stdout.splitlines():
+            m = url_r.match(line)
+            if m:
+                urls_by_name[m.group(1)] = m.group(2)
+
+        if len(paths_by_name) != len(urls_by_name):
+            log.error("Error matching paths with urls in {}/{}"
+                      .format(self.name, GitRepo.SUBMODULE_FILE))
+            sys.exit(1)
+
+        submodules = []
+        for name_key, path in paths_by_name.items():
+            # We use the relative path within the repository to ask the git index
+            # for the pointer that's currently commited
+            submodule_ref = self._get_submodule_pointer(revision, path)
+
+            url = urls_by_name[name_key]
+            name = name_key
+            if "/" in name:
+                # Wit only supports a 'flat' checkout pattern.
+                # By default, git submodules uses the relative checkout path to name the submodule.
+                # The user can add an explict name, but often do not.
+                # So if the submodule happens to have a name that looks like a nested path,
+                # use the the final path component in the remote url instead as the path-based names
+                # as sometimes they're easy-to-clash names like docs/html.
+                name = os.path.basename(url).replace(".git", "")
+
+            submodules.append(RepoEntry(name, submodule_ref, url))
+
+        return submodules
+
+    def _get_submodule_pointer(self, revision, path):
+        """
+        Get the submodule pointer commit in the index.
+        Note: This is NOT always the currently checked-out commit for a submodule.
+        The 'git ls-tree' output is defined as:
+           ^<mode> space <type> space <object> tab <file>$
```

Comment:
```suggestion
           ^<mode> space <type> space <git object hash> tab <file>$
```

I think explicitly calling this the "git object hash" will make this a bit more understandable by people who are less familiar with deeper git internals.